mono/utils/mono-threads-coop.c

```diff
@@ -197,6 +198,53 @@ copy_stack_data (MonoThreadInfo *info, MonoStackData *stackdata_begin)
 	state->gc_stackdata_size = stackdata_size;
 }
 
+#ifdef _MSC_VER
+typedef void (*CopyStackDataFunc)(MonoThreadInfo *, MonoStackData *, gconstpointer, gconstpointer);
+
+#ifdef TARGET_AMD64
+typedef struct {
+	__m128d fregs [10];
+	host_mgreg_t gregs [8];
+} MonoBuiltinUnwindInfo;
+
+// Defined in win64.asm
+G_EXTERN_C void
+copy_stack_data_internal_win32_wrapper (MonoThreadInfo *, MonoStackData *, MonoBuiltinUnwindInfo *, CopyStackDataFunc);
+#else
+typedef struct {
+	host_mgreg_t gregs [4];
+} MonoBuiltinUnwindInfo;
+
+__declspec(naked) void
```

Comment:
Maybe an explicit `__cdecl` or `__fastcall` or `__stdcall`? Or maybe the compiler default __cdecl is inescapable w/o massive code churn. Just that most code is calling convention-agnostic, and this probably is not. I admit I'm not reading it closely yet.

Reply:
Might be nice to use jmp_buf and setjmp instead. Might.
Or explicitly list what is volatile/non-volatile on NT/x86.
That documentation is difficult to find.

Reply:
The dumped registers are inline with what you get on mingw gcc x86 implementation of __builtin_unwind_init, that's the only documentation that I could find regarding that.

Reply:
x86

https://docs.microsoft.com/en-us/cpp/cpp/argument-passing-and-naming-conventions?view=vs-2019


The compiler generates prolog and epilog code to save and restore the ESI, EDI, EBX, and EBP registers, if they are used in the function.

So those are the non-volatiles.
And it matches your code.

Reply:
And amd64:

https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019

The registers RBX, RBP, RDI, RSI, RSP, R12, R13, R14, and R15 are considered nonvolatile and must be saved and restored by a function that uses them.

which matches your code, except for rsp, which is maybe special.

Reply:
Jepp, all inline with __builtin_unwind_init implementation on Windows x86/x64.